### PR TITLE
[PE-46] fix: added aspect ratio to resizing

### PR DESCRIPTION
**Pull request by Palanikannan1437**
## Description

1. Added aspect ratio to image attributes and I'm no more setting the height on images
2. Added a smooth blur effect and loading state for images while being uploaded
3. Added strict types for Pixel values
4. Restricted the click event on the editor container only to the clicks by the user specifically on the editor container at the bottom by adding padding to it.
5. Fixed the image insertion position via the Toolbar
6. Handled the case where image is of old type with width and height attributes are numbers and not pixel value
7. Handled a case where image doesn't load, showing an error state
8. Fixed the image resize handler event listener logic to be added only while resizing making it really smooth.
9. Refactored the entire code for having a better separation between the upload phase and rendering the final image phase.
10. Added the logic to not allow closing modals while uploading images.
11. Added the trimming of large file names logic before uploading.
12. Handled various loading states across the lifecycle of the image upload minimizing layout shifts. 
13. Focusing out of the image carefully depending on what's the current node in selection and the node below it.
14. Inserting multiple images (while dropping/pasting) carefully taking the last node insertion's status to avoid range errors while inserting.

## Release Notes

- **New Features**
  - Enhanced image upload functionality with improved state management and visual feedback for upload failures.
  - Optimized performance for image handling through memoization and refined file upload processes.

- **Bug Fixes**
  - Improved error handling during image uploads and loading states.

- **Style**
  - Adjusted styling conditions to reflect new image loading states.

- **Documentation**
  - Updated internal documentation to reflect changes in image upload and handling functionalities.

### CODE REVIEW

File: packages/editor/src/core/extensions/custom-image/components/image-block.tsx [hunk 1]
```diff
@@ -115,65 +177,81 @@ export const CustomImageBlock: React.FC<CustomImageNodeViewProps> = (props) => {
...
-    editorContainer.addEventListener("mouseup", handleMouseUp);
-    editorContainer.addEventListener("mouseleave", handleMouseLeave);
-
-    return () => {
-      editorContainer.removeEventListener("mousemove", handleMouseMove);
-      editorContainer.removeEventListener("mouseup", handleMouseUp);
-      editorContainer.removeEventListener("mouseleave", handleMouseLeave);
-    };
-  }, [handleResize, handleResizeEnd, editorContainer]);
+  const showImageLoader = !(remoteImageSrc || imageFromFileSystem) || failedToLoadImage || !initialResizeComplete;
```
Comment (rahulramesha) [inline packages/editor/src/core/extensions/custom-image/components/image-block.tsx:184]: This looks a bit confusing with so many checks, can you please try to simplify this or add comments to this

File: packages/editor/src/core/extensions/custom-image/components/image-block.tsx [hunk 2]
```diff
@@ -115,65 +177,81 @@ export const CustomImageBlock: React.FC<CustomImageNodeViewProps> = (props) => {
...
+  const showImageUtils = editor.isEditable && remoteImageSrc && !failedToLoadImage && initialResizeComplete;
```
Comment (rahulramesha) [inline packages/editor/src/core/extensions/custom-image/components/image-block.tsx:185]: This As well

File: packages/editor/src/core/extensions/custom-image/components/image-uploader.tsx [hunk 3]
```diff
@@ -44,42 +119,38 @@ export const CustomImageUploader = (props: {
...
       onDragOver={onDragEnter}
       onDragLeave={onDragLeave}
       contentEditable={false}
-      onClick={handleUploadClick}
+      onClick={() => fileInputRef.current?.click()}
     >
       <ImageIcon className="size-4" />
       <div className="text-base font-medium">
-        {loading ? "Uploading..." : draggedInside ? "Drop image here" : existingFile ? "Uploading..." : "Add an image"}
+        {failedToLoadImage
```
Comment (rahulramesha) [inline packages/editor/src/core/extensions/custom-image/components/image-uploader.tsx:141]: please avoid this nested ternary logic. simplify it with if else and comments. Also move it out to a different function if possible

File: packages/editor/src/core/extensions/custom-image/components/image-uploader.tsx [hunk 4]
```diff
@@ -41,45 +116,41 @@ export const CustomImageUploader = (props: {
...
+          ? "Error loading image"
+          : isImageBeingUploaded
+            ? "Uploading..."
+            : draggedInside
+              ? "Drop image here"
+              : existingFile
+                ? "Uploading..."
+                : "Add an image"}
```
[the remark this answers is not in the document]
Comment (rahulramesha) [reply packages/editor/src/core/extensions/custom-image/components/image-uploader.tsx:149]: check this Refactor as well

File: packages/editor/src/core/extensions/custom-image/components/image-block.tsx (hunk 1 above)
Comment (Palanikannan1437) [reply packages/editor/src/core/extensions/custom-image/components/image-block.tsx:184]: Yup will do that

File: packages/editor/src/core/extensions/custom-image/components/image-uploader.tsx (hunk 3 above)
Comment (Palanikannan1437) [reply packages/editor/src/core/extensions/custom-image/components/image-uploader.tsx]: Sure thing, refactored this onto a function.

**Pull request merged**